Add Execute flow and DeviceManager implementation

cdaly333 opened the pull request:
This PR adds support for testing a fulfillment's `executeHandler`.  This support is added through implementation of `DeviceManagerStub`, which responds to pre-defined commands with pre-defined responses.  Also includes various testing tweaks.

**Changes**
- `DeviceManagerStub` was fully implemented, minus `getProxyInfo()` (looking into this)
- A `triggerExecute()` function was added to `MockLocalHomePlatform` to communicate with a fulfillment's `executeHandler`
- Stubs for `smarthome.Execute`, `smarthome.Constants.Protocol`
- Module-level function to access `DeviceManagerStub` from an `App`
- Tests for Execute functionality in `test/platform/test-platform-execute.ts`
- Tests for `DeviceManagerStub` in `test/example/device-manager.ts`
- Refactor existing tests with new fixtures and Execute requirements
- A fixtures file, `test/platform/test-platform-fixtures` was added for internal testing helper functions and constants
- JSDocs for all new functions.  Formatted and tweaked some old comments


**Issues:**
- Resolves #16 
- Resolves #25 
- Resolves #23 
- Resolves #4

Tagging @limlam

proppy commented on test/example/test.ts:
```diff
@@ -69,7 +69,10 @@ test('identify-handler-registers-local-id', async t => {
   await t.notThrowsAsync(async () => {
     t.is(
       // This call will return the local device id
-      await mockLocalHomePlatform.triggerIdentify(discoveryBuffer),
```
can you add a test for execute?

cdaly333 replied:
Added a simple success path test

Fleker commented on test/platform/test-platform-fixtures.ts:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Fixtures used across internal tests
+ */
+export function makeSendCommand(
```
Functions that are exported should have some JSDoc.

cdaly333 replied:
Done

Fleker commented on test/platform/test-platform-identify.ts:
```diff
@@ -80,23 +82,24 @@ test('trigger-identify-with-undefined-verificationId-throws', async t => {
+    ),
+    deviceManager
+  );
+  app.onIdentify(validIdentifyHandler).onExecute(validExecuteHandler).listen();
```
Is there an example of using an `invalidExecuteHandler`?

cdaly333 replied:
Added invalid examples in `execute-handler-sends-wrong-buffer`.  Renamed this local variable to `executeHandler`

cdaly333 commented on src/platform/device-manager.ts:
```diff
@@ -3,23 +3,75 @@
+  /**
+   * Checks if the next request marked pending matches a given request.
+   * @param requestToMatch  A request to test against the next request marked pending.
+   * @returns  Promise that resolves to a boolean representing if the requests matched.
```
It will hang if `markPending()` is never called, but will resolve with `false` if there was a call without a match.  `ava` detects hanging promises with a `Promise returned by test never resolved` message.  It would be better if I could explicitly reject it, but I'm not sure at what point I should deem the `Promise` rejected / timed out.

Fleker replied:
If `pendingRequest === requestToMatch` is `false`, should the promise reject?

cdaly333 replied:
Done

cdaly333 commented on src/platform/stub-setup.ts:
```diff
@@ -4,25 +4,108 @@
+  app: smarthome.App
+): DeviceManagerStub {
+  const deviceManager = app.getDeviceManager();
+  if (deviceManager instanceof DeviceManagerStub) {
```
There's no chance this won't be true.  We could remove these functions, but then the developer would need to manually downcast their `App` to `AppStub` in every test and mute the `eslint` error every time they want to use the stubs.  These functions also help hide `AppStub` from ever being used in a developer test.

The other option is to require the developer to use explicitly construct `AppStub` objects.  They would then have direct access to the stubs.  This is probably what I'll end up doing, since the developer is using other stubs explicitly as well.

@proppy also suggested removing these functions.

proppy replied:
maybe you could provide both with a single top-module function that return an interface?

cdaly333 replied:
Done

cdaly333 commented on src/platform/device-manager.ts:
```diff
@@ -3,23 +3,75 @@
+  }
+
+  /**
+   * Registers a command that will be checked on `send()` anCommd a corresponding
```
Done.

cdaly333 commented on src/platform/mock-local-home-platform.ts:
```diff
@@ -86,4 +94,63 @@ export class MockLocalHomePlatform {
+    }
+
+    const executeRequest: smarthome.IntentFlow.ExecuteRequest = {
+      requestId: requestId,
```
Done

cdaly333 commented on src/platform/mock-local-home-platform.ts:
```diff
@@ -86,4 +94,63 @@ export class MockLocalHomePlatform {
+                ],
+              },
+            ],
+            structureData: {},
```
I believe so.  It's `@hidden` and omitted from all documentation.  I'm assuming it's used internally.

Fleker replied:
If it's not in the documentation and is hidden, should we be sending it?

cdaly333 replied:
It's required in order to form an `IdentifyRequest`, since the `Payload<LocalIdentifiedDevice>` type requires it.  I could stub `IdentifyRequest` if we want to remove it.

cdaly333 commented on src/platform/mock-local-home-platform.ts:
```diff
@@ -86,4 +94,63 @@ export class MockLocalHomePlatform {
+
+    const response = await this.app.executeHandler(executeRequest);
+
+    // Only valid command length supported is 1
```
I changed this function to take in a list of `ExecuteResponseCommands` and forward it directly to the fulfillment.

cdaly333 commented on src/platform/mock-local-home-platform.ts:
```diff
@@ -86,4 +94,63 @@ export class MockLocalHomePlatform {
+   * @param deviceId  The device id to set in the `ExecuteRequest`
+   * @param command  The command string to set as the single command in the `ExecuteRequest`
+   * @param params  The execution parameters to set in the single command in the `ExecuteRequest`
+   */
```
Done

cdaly333 commented on src/platform/stub-setup.ts:
```diff
@@ -4,25 +4,108 @@
   },
+  DataFlow: {
+    UdpRequestData: class {
+      data!: string;
```
Replaced with default values

cdaly333 commented on src/platform/stub-setup.ts:
```diff
@@ -4,25 +4,108 @@
+  },
+  Execute: {
+    Response: {
+      Builder: class {
```
Moved to external variable.

cdaly333 commented:
- changed `triggerExecute` to take in raw `ExecuteRequestCommands`.
- moved some stubs around
- added example execute integration test
- added unit tests for `DeviceManagerStub` class

Still need to address a few comments

proppy commented on test/example/execute-test.ts:
```diff
@@ -0,0 +1,94 @@
+const EXECUTE_REQUEST_ID = 'request-id-123';
+const app: smarthome.App = new smarthome.App('0.0.1');
+
+function createSimpleExecuteCommands(
```
should this helper live into the library?

cdaly333 replied:
Done

proppy commented on test/example/identify-test.ts:
```diff
@@ -0,0 +1,49 @@
+/*
```
what about consolidating both examples test in one file?

cdaly333 replied:
I may combine them later, but I think I'll be adding more tests to `execute-test.ts` so just want to make sure the combined file won't be too big

proppy commented on test/example/execute-test.ts:
```diff
@@ -0,0 +1,94 @@
+  const executeCommands = createSimpleExecuteCommands(
+    DEVICE_ID,
+    'actions.devices.commands.OnOff',
+    {}
```
no params?

cdaly333 replied:
Added simple params

proppy commented on test/example/execute-test.ts:
```diff
@@ -0,0 +1,94 @@
+): smarthome.IntentFlow.ExecuteRequestCommands {
+  return {
+    devices: [{id: deviceId}],
+    execution: [{command, params}],
```
most command would rely on customData for dispatching the command should we show an example for that>

cdaly333 replied:
I added a `customData` parameter and value for this helper, and changed the example test to use a simple `customData` object.

cdaly333 commented on src/platform/device-manager.ts:
```diff
@@ -3,23 +3,81 @@
+   * @param expectedCommand  The command to check against incoming commands.
+   * @param response  The response to send when an incoming command matches.
+   */
+  public addExpectedCommand(
```
I'm not really sure what you mean.  If we buffer the `DeviceManager` commands and verify them after they are sent, how would we mock the responses per-command? (Developers may want to test different scenarios with `CommandSuccess` and `CommandFailue` responses depending on the command)

cdaly333 replied:
I guess it might help is to buffer the sent commands anyways.  In case a command doesn't match, it would allow the developer to see what commands were sent vs. what was expected.

cdaly333 replied:
I added a buffer to store sent commands and some assertion helpers.  I updated the tests to use them.

Fleker commented on src/platform/execute.ts:
```diff
@@ -0,0 +1,75 @@
+ * @param command  The single command to send to the device
+ * @param params  Parmeters for the command
+ */
+export function createSimpleExecuteCommands(
```
Should add a `@returns`

cdaly333 replied:
Done

Fleker commented on test/platform/test-platform-fixtures.ts:
```diff
@@ -1,51 +1,15 @@
+ * Creates a basic Identify handler that returns the specified id and verificationId.
+ * @param deviceId  The `id` to include in the `IdentifyResponse` in the handler.
+ * @param verificationId  The `verificationId` to include in the `IdentifyResponse` in the handler.
+ */
```
Be sure to add an `@returns`

cdaly333 replied:
Done

Fleker commented on src/platform/stub-setup.ts:
```diff
@@ -5,31 +5,50 @@
+      constructor(requestId: string, errorCode?: string, debugString?: string) {
+        super(errorCode);
+        this.requestId = requestId;
+        this.errorCode = errorCode || ' ';
```
Does this empty string need to have a space?

cdaly333 replied:
No, it should be empty.  Fixed

Fleker commented on test/platform/test-device-manager.ts:
```diff
@@ -0,0 +1,125 @@
+ */
+test('device-manager-expected-mark-pending', async t => {
+  const deviceManager = new DeviceManagerStub();
+  const executeReqeust = createExecuteRequest('action.devices.commands.OnOff');
```
`executeRequest`

Fleker replied:
Don't we need parameters along with a command?

cdaly333 replied:
Since this test is just differentiating any two non-identical execute requests, it's not really required here, but I added it for completion

Fleker commented on test/platform/test-device-manager.ts:
```diff
@@ -0,0 +1,125 @@
+ */
+test('device-manager-unexpected-mark-pending', async t => {
+  const deviceManager = new DeviceManagerStub();
+  const executeReqeust = createExecuteRequest('action.devices.commands.OnOff');
```
`executeRequest`

cdaly333 replied:
Done

Fleker commented on test/platform/test-platform-execute.ts:
```diff
@@ -2,74 +2,91 @@
+    Buffer.from('test-buffer'),
+    deviceId
+  ),
+    localDeviceId;
```
Is this value indented correctly? I'm not sure I follow where this gets called.

cdaly333 replied:
Huh, it actually doesn't belong there at all.  I removed it.  I'm surprised it compiled.

Fleker commented on test/platform/test-platform-fixtures.ts:
```diff
@@ -54,32 +18,9 @@ export function createIdentifyHandler(
         device: {
           id: deviceId,
-          verificationId: veriicationId,
+          verificationId: verificationId,
```
As the key and variable are the same, you can just do `verificationId,` like in other places.

cdaly333 replied:
Done

Fleker commented on test/platform/test-platform-identify.ts:
```diff
@@ -96,7 +90,7 @@ test('trigger-identify-with-valid-state', async t => {
   );
   app.onIdentify(validIdentifyHandler).onExecute(validExecuteHandler).listen();
-  const mockLocalHomePlatform = extractMockLocalHomePlatform(app);
+  const mockLocalHomePlatform = extractStubs(app).mockLocalHomePlatform;
```
You can do destructuring `const {mockLocalHomePlatform} = extractStubs(app)`

cdaly333 replied:
Neat. Done

Fleker commented on test/platform/test-platform-identify.ts:
```diff
@@ -59,11 +57,8 @@ test('trigger-identify-with-undefined-verificationId-throws', async t => {
-    .listen();
-  const mockLocalHomePlatform = extractMockLocalHomePlatform(app);
+  app.onIdentify(invalidIdentifyHandler).onExecute(executeHandler).listen();
+  const mockLocalHomePlatform = extractStubs(app).mockLocalHomePlatform;
```
You can do destructuring `const {mockLocalHomePlatform} = extractStubs(app)`

cdaly333 replied:
Done

Fleker commented:
I'm pretty happy with this, not sure if Proppy wants to give any more input
